pkg/autodiscovery/listeners/snmp.go

```diff
@@ -261,6 +274,19 @@ func (l *SNMPListener) checkDevices() {
 	}
 }
 
+func (l *SNMPListener) collectMetrics(subnets []snmpSubnet) {
+	sender, senderErr := aggregator.GetSender(senderCheckID)
+	if senderErr != nil {
+		log.Errorf("Error getting aggregator sender: %s", senderErr)
+		return
+	}
+	for _, subnet := range subnets {
+		tags := []string{"network:" + subnet.configNetwork}
+		sender.Gauge("snmp.discovered_devices_count", float64(len(subnet.devices)), "", tags)
```

Comment:
can we move this metrics in the check himself? 

it looks strange to me to have two source for the same metrics "snmp" namespace: the check and and listener.

Reply:
@clamoriniere 
> can we move this metrics in the check himself?
Not really possible, since only `snmp` listener know how many devices has been discovered. When using snmp listener, the snmp integrations checks only collect metrics for individual devices.

> it looks strange to me to have two source for the same metrics "snmp" namespace: the check and and listener.

About the metric naming, it could be called something different, but using `snmp.discovered_devices_count` is for keeping the same naming as the equivalent metric sent from the autodiscovery feature of snmp check itself. (We can possibly document this part to avoid confusion).

Reply:
Moving the metric to the Python check (not sure how though) would probably also result in sending duplicate copies of this metric (once per instance in a simple scenario where we don't require users to mark an instance as "this one sends discovery metrics", which looks awkward to me too).

Reply:
* what's the purpose of this metric exactly?
* is this a metric users will actively make use of?
* how often is the logic that sends this metric called?

FYI, only checks are supposed to use the aggregator this way. Using the aggregator from AD listeners may work today, but isn't really part of the aggregator's design so I don't think it should be expected to work in the future.

For context: the agent exposes internal metrics using go expvar (widespread in agent codebase) and openmetrics/prometheus (less widespread, but allows adding tags to the metrics). These can be reported to DD using the expvar check and openmetrics checks (with some custom config), and are sent with the `datadog.agent.` prefix.
At the moment, there's no standard way to report metrics (that end users are expected to use on DD) from the Agent but outside of the checks. If this is needed I don't think we should use `aggregator.Sender` as-is for that.

Reply:
@olivielpeau 

> what's the purpose of this metric exactly?

The metric `snmp.discovered_devices_count` is to report the number of devices discovered by the autodiscovery listener, it's equivalent to the one we send in snmp python autodiscovery: https://github.com/DataDog/integrations-core/blob/6ec9d74ee303dc54e4b171a2b0ab274c28dac39b/snmp/datadog_checks/snmp/snmp.py#L340

> is this a metric users will actively make use of?

I think so, the metrics helps making sure we are getting the right amount of devices by the autodiscovery mechanism.

> how often is the logic that sends this metric called?

It's called every time the auto discovery is runs, it can be configured via `snmp_listener.discovery_interval`, by default 3600 secs.

> FYI, only checks are supposed to use the aggregator this way. Using the aggregator from AD listeners may work today, but isn't really part of the aggregator's design so I don't think it should be expected to work in the future.

Ok, using ` aggregator.GetSender` like here is probably not the right to do it then.

> For context: the agent exposes internal metrics using go expvar (widespread in agent codebase) and openmetrics/prometheus (less widespread, but allows adding tags to the metrics). These can be reported to DD using the expvar check and openmetrics checks (with some custom config), and are sent with the datadog.agent. prefix.

I see. It might not be optimal from user to have to setup an expvar check or openmetrics check to get one metric for snmp autodiscovery.

> At the moment, there's no standard way to report metrics (that end users are expected to use on DD) from the Agent but outside of the checks.

Do you think we can plan for that ? Not sure if a "minor" use case like this one worth the effort and get it prioritised though.

--- 

That said, there is an alternative way to count the devices, that is to send a count of 1 for each device from `snmp integration` (instead of autodiscovery as in this PR) and sum that from DD site. e.g. `snmp.devices_count` tagged by device ip and network.